core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -177,6 +177,8 @@ private[spark] class DAGScheduler(
   // TODO: Garbage collect information about failure epochs when we know there are no more
   //       stray messages to detect.
   private val failedEpoch = new HashMap[String, Long]
+  // In addition, track epoch for failed executors that result in lost file output
```

Comment:
I'll try to expand on this comment.

Reply:
yeah I really agree with this.  I think we need to rename both of these epochs.  I think I finally get why they both matter. One tracks files owned by the external shuffle service -- so you don't update those files when you learn that an executor is lost.  The other tracks files owned by the executor itself.  So that includes memory-cached blocks always, and also shuffle files w/out the external shuffle service.

Once I wrapped my head around that part, the change in the logic made sense.

I will think some more about better names and comments which would help with that part

Reply:
how about `shuffleFileLostEpoch` here, and a comment building on the suggested comment above:

Tracks the latest epoch of a fully processed error which indicated shuffle files have been lost from the given executor.

This is closely related to `executorFailureEpoch`.  In fact, anytime an entry is added here, it must also be added to `executorFailureEpoch`, so `shuffleFileLostEpoch(exec) <= executorFailureEpoch(exec)`.  These only differ when the external shuffle service is enabled.  In that case, when an executor is lost, we do *not* update the epoch here; we wait for a fetch failure.  This way, if only the executor fails, we do not remove the shuffle data as it can still be served by the shuffle service; but if there is a failure in the shuffle service, we remove the shuffle data only once, though we are likely to get many failures.

Reply:
I changed `fileLostEpoch` to `shuffleFileLostEpoch` and more or less adopted your suggestion for the comment explaining it.